deep_qa/data/dataset_readers/language_modeling.py

```diff
@@ -0,0 +1,111 @@
+        one with default parameters.
+    """
+    def __init__(self,
+                 filename: str,
+                 tokens_per_instance: int=None,
+                 tokenizer: Tokenizer=WordTokenizer(),
+                 token_indexers: List[TokenIndexer]=None):
+        self._filename = filename
+        self._tokens_per_instance = tokens_per_instance
+        self._tokenizer = tokenizer
+        if token_indexers is None:
+            token_indexers = [SingleIdTokenIndexer()]
+        self._token_indexers = token_indexers
+        self._start_token = '<S>'
+        self._end_token = '</S>'
+
+    @overrides
+    def read(self):
+        with open(self._filename, "r") as text_file:
+            instance_strings = text_file.readlines()
+        if self._tokens_per_instance is not None:
+            all_text = " ".join([x.replace("\n", " ").strip() for x in instance_strings])
+            tokenized_text = self._tokenizer.tokenize(all_text)
+            num_tokens = self._tokens_per_instance
+            tokenized_strings = []
+            for index in range(0, len(tokenized_text) - num_tokens, num_tokens):
+                tokenized_strings.append(tokenized_text[index:index + num_tokens])
+        else:
+            tokenized_strings = [self._tokenizer.tokenize(s) for s in instance_strings]
+
+        # TODO(matt): this isn't quite right, because you really want to split on sentences,
+        # tokenize the sentences, add the start and end tokens per sentence, then change the tokens
+        # per instance if desired.  But, we can fix that later, if someone actually wants to use
+        # this for language modeling.  This is just another example of how to use the data reader
+        # code, for now.
+        tokenized_strings = [[self._start_token] + x + [self._end_token] for x in tokenized_strings]
+
+        # No matter how you want to represent the input, we'll always represent the output as a
+        # single token id.  This code lets you learn a language model that concatenates word
+        # embeddings with character-level encoders, in order to predict the word token that comes
+        # next.
+        output_indexer = None
+        for indexer in self._token_indexers:
+            if isinstance(indexer, SingleIdTokenIndexer):
+                output_indexer = indexer
+                break
+        else:
+            output_indexer = SingleIdTokenIndexer()
+
+        instances = []
+        for tokenized_string in tokenized_string:
+            input_field = TokenizedTextField(tokenized_string[:-1], self._token_indexers)
+            output_field = TokenizedTextField(tokenized_string[1:], [output_indexer])
+            instances.append(Instance({'input_tokens': input_field,
+                                       'output_tokens': output_field}))
+        return Dataset(instances)
+
+    @classmethod
+    def from_params(cls, params: Params):
```

Comment:
Ok fine on the not doing the expensive thing in the init call, but still why doesn't the signature look like this:
```
dataset_reader_params = params.pop('dataset')
dataset_type = concrete_dataset_readers[dataset_reader_params.pop("type")]

dataset_reader = dataset_type(dataset_reader_params)
dataset = dataset_reader.read()  # expensive operation
```
The classmethod on each DatasetReader is different so we still need to know which dataset_reader we are using beforehand - it seems like having this parameter unpacking inside the `DatasetReader.__init__` is effectively equivalent?